### DISCUSSION TITLE
Reuse folder tree view in extension or add folders to Explorer view

### DISCUSSION THREAD

**Discussion by itavero** [microsoft/vscode-discussions#571] [Extension Development QnA] [answered]
I'm thinking about creating an extension to accompany the way we reuse code across projects (which uses CPM.cmake).
For this I would like to be able to easily view the source files of the packages that are included, so I was thinking about adding a tree view to the Explorer view that lists all the packages that are added and would allow you to browse through the files.

Is there a way to either reuse the folder tree view that is already used for workspace folders in the Explorer?
Or can I perhaps add a folder there?

I did notice that it is possible to add workspace folders, but then this would probably also load extensions for that folder, which is probably not what I want as I expect that this will lead to issues.
It would however be nice to have the right file icons and such show up in the tree view as well as having the context menu that is normally present for files/folders.

(The way CPM.cmake roughly works is that it has CMake checkout repositoires to a cache directory and then includes those projects in the CMake project. In other words, they are already loaded in my "main" workspace CMake project and I don't care for loading them multiple times.)

**Comment by r3m0t** [comment, marked as answer]
It's better UX to contribute a separate tree view that shows in the Explorer view container. Like this example that shows the dependencies in a JS project. https://code.visualstudio.com/api/extension-guides/tree-view

To display in the Explorer, you must have a filesystem (which could be a virtual filesystem) and use `updateWorkspaceFolders` to add a path from the virtual filesystem into the workspace. Like this example. You can modify it to add the folders to the workspace without deleting the existing folders. https://github.com/Microsoft/vscode-extension-samples/tree/main/fsprovider-sample

If you pass a file URI to `updateWorkspaceFolders`, it's much like the user doing `File > Add Folder to Workspace`.

And you could set `files.exclude` to hide the duplicate files from the "real" folders.

**Reply by itavero** [reply]
I also think it is better to use a separate tree view, but how do I prevent duplicating all the code to get the right file icons , the file context menu, and such?

**Reply by r3m0t** [reply]
File icons - set `TreeItem.resourceUri`

Context menu - no real way, but your extension can call the built in commands for each item so you don't need to reimplement the actual functionality. 

I would try both methods and see the UX and decide which way to go based on that.
